common/endian.h

```diff
@@ -239,14 +239,10 @@
  *  @{
  */
 
-// Test for GCC >= 4.0. These implementations will automatically use
+// Test for GCC and compatible. These implementations will automatically use
 // CPU-specific instructions for unaligned data when they are available (eg.
-// MIPS). See also this email thread on scummvm-devel for details:
-// <http://thread.gmane.org/gmane.games.devel.scummvm/8063>
-//
-// Moreover, we activate this code for GCC >= 3.3 but *only* if unaligned access
-// is allowed.
-#if GCC_ATLEAST(4, 0) || (GCC_ATLEAST(3, 3) && !defined(SCUMM_NEED_ALIGNMENT))
+// MIPS).
+#if defined(__GNUC__)
```

Comment:
you missed `!defined(SCUMM_NEED_ALIGNMENT)`

Reply:
@lephilousophe I believe that `!defined(SCUMM_NEED_ALIGNMENT)` was just for GCC > 3.3 and < 4.0 (which had no C++11 support).

So, here, all supported clang/gcc versions would always return true for the first `#if GCC_ATLEAST(4, 0)` case, and so the second case will never be reached anyway, so we can just keep the first condition and simplify it to `#if defined(__GNUC__)`, no?

(Well, I may have got my logic wrong somewhere, though :)

Reply:
`SCUMM_NEED_ALIGNMENT` is defined when platform doesn't support unaligned memory accesses (DS, Dreamcast, ...) in `common/scummsys.h` or in `configure` script.
The code protected by this `if defined` heavily relies on not portable code and would crash if enabled on one of these platforms while doing unaligned memory access.
It really should stay here.

Reply:
But most strict-alignment archs were already returning true on the first `#if GCC_ATLEAST(4, 0)` condition for years, no? So the second part wasn't even evaluated (except for the strict-alignment archs using GCC < 4.0), no?

The original code was this:

```c++
#if GCC_ATLEAST(4, 0) || (GCC_ATLEAST(3, 3) && !defined(SCUMM_NEED_ALIGNMENT))
// GCC >= 4.0 evaluated first, so SCUMM_NEED_ALIGNMENT is only tied to the GCC >= 3.3 && GCC < 4.0 case
```

not this:

```c++
#if (GCC_ATLEAST(4, 0) || GCC_ATLEAST(3, 3)) && !defined(SCUMM_NEED_ALIGNMENT)
// SCUMM_NEED_ALIGNMENT checked for any GCC version because of different precedence rules
```

Here's some context when that line was added:
https://scummvm-devel.narkive.com/5nJgFIyc/wii-gamecube-resp-powerpc-vs-unaligned-memory-access#post16

> [...] *starting with gcc 4.0, the code generators should output CPU specific instructions for unaligned data with the code used in that path. Since the idea behind that code path was back then to use the (probably) more efficient CPU specific instructions it was only limited to alignment requiring CPU types.*

So it looks like the `SCUMM_NEED_ALIGNMENT` check in that particular place was for performance reasons, not portability reasons? Isn't `__attribute__ ((__packed__)` what is supposed to make it work on strict-alignment too?

Anyway, I don't mind adding this `!defined(SCUMM_NEED_ALIGNMENT)` check, but then it would mean that there was a long-standing bug on strict-alignment archs.

Reply:
Indeed you are right I misread the #if condition.
So recent enough GCC should handle it properly in every architectures.
The `__packed__` and `__may_alias__` should force the compiler to emit correct code.

I wonder if another way to do it could be to simply use memcpy but that's another story.